provider/packet/environ.go

```diff
@@ -0,0 +1,322 @@
+	environscloudspec "github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/instances"
+	envinstance "github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/provider/common"
+	"github.com/juju/juju/storage"
+	"github.com/juju/schema"
+	"github.com/juju/utils/arch"
+	"github.com/juju/version"
+	"github.com/packethost/packngo"
+	"gopkg.in/juju/environschema.v1"
+)
+
+type environConfig struct {
+	config *config.Config
+	attrs  map[string]interface{}
+}
+
+type environ struct {
+	ecfgMutex    sync.Mutex
+	ecfg         *environConfig
+	name         string
+	cloud        environscloudspec.CloudSpec
+	packetClient *packngo.Client
+}
+
+var providerInstance environProvider
+
+func (e *environ) AdoptResources(ctx context.ProviderCallContext, controllerUUID string, fromVersion version.Number) error {
+	trace()
+	return nil
+}
+
+func (e *environ) Bootstrap(ctx environs.BootstrapContext, callCtx context.ProviderCallContext, args environs.BootstrapParams) (*environs.BootstrapResult, error) {
+	trace()
+	return common.Bootstrap(ctx, e, callCtx, args)
+}
+
+func (e *environ) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag("juju-instance", "")
+}
+
+// if values tag and state are left empty it will return all instances
+func (e *environ) getPacketInstancesByTag(tag, state string) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+	opt := &packngo.ListOptions{Search: tag}
+
+	devices, _, err := e.packetClient.Devices.List(e.cloud.Credential.Attributes()["project-id"], opt)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, d := range devices {
+		if state != "" && d.State == state {
+			toReturn = append(toReturn, &packetDevice{e, &d})
+		} else if state == "" {
+			toReturn = append(toReturn, &packetDevice{e, &d})
+		}
+	}
+
+	return toReturn, nil
+}
+
+func (e *environ) AllRunningInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag("juju-instance", "active")
+}
+
+func (e *environ) Config() *config.Config {
+	e.ecfgMutex.Lock()
+	defer e.ecfgMutex.Unlock()
+	return e.ecfg.config
+}
+
+func (e *environ) ConstraintsValidator(ctx context.ProviderCallContext) (constraints.Validator, error) {
+	validator := constraints.NewValidator()
+	validator.RegisterUnsupported([]string{constraints.CpuPower, constraints.VirtType})
+	validator.RegisterConflicts([]string{constraints.InstanceType}, []string{constraints.Mem})
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL})
+	return validator, nil
+}
+
+func (e *environ) ControllerInstances(ctx context.ProviderCallContext, controllerUUID string) ([]instance.Id, error) {
+	trace()
+	return nil, nil
+}
+
+func (e *environ) Create(ctx context.ProviderCallContext, args environs.CreateParams) error {
+	trace()
+	return nil
```

Comment:
Do I need to implement this function?

Reply:
I see that the other providers do implement this. It could be something as simple as verifying that the provided credentials work (for example, the ec2 implementation checks the credentials and then proceeds to validate VPC details) so you can fail early if the credentials are wrong/disabled etc.